autogo/def.bzl

```diff
@@ -0,0 +1,99 @@
+  mach = _exec(repo_ctx, ["uname", "-m"], quiet=True).stdout.strip()
+  if mach != "x86_64":
+    fail("autogo only supports x86_64, not %s" % mach)
+  repo_ctx.execute(["uname", "-m"])
+  if repo_ctx.os.name == "mac os x":
+    return mac
+  elif repo_ctx.os.name == "linux":
+    return linux
+  fail("autogo only supports linux and mac/darwin, not %s" % repo_ctx.os.name)
+
+
+def _impl(repo_ctx):
+  print("Generating rules for %s..." % repo_ctx.attr.prefix)
+  # tell gazelle which package it is generating
+  prefix = repo_ctx.attr.prefix
+  # find the autogo binary for the host env
+  autogo = repo_ctx.path(_select(repo_ctx, repo_ctx.attr._autogo_linux, repo_ctx.attr._autogo_darwin))
+  # find the gazelle binary for the host env
+  gazelle = repo_ctx.path(_select(repo_ctx, repo_ctx.attr._gazelle_linux, repo_ctx.attr._gazelle_darwin))
+  # get the path to the root workspace
+  root_workspace = repo_ctx.path(repo_ctx.attr._workspace).dirname
+
+  # Create a shadow copy of root_workspace in the repo workspace
+
+  # First, delete anything we created last time
+  # TODO(fejta): massively unoptimized, but expedient way to ensure correctness
+  _exec(repo_ctx, ["rm", "-rf", repo_ctx.path(".")])
+  _exec(repo_ctx, ["touch", repo_ctx.path("BUILD.bazel")]) # required by bazel
+  # Update the timestamp of the autogo binary, invalidating the autogo result cache.
+  # Critically, this ensures that we reevaluate the root workspace for any changes.
+  # Otherwise, if a file/package is added/deleted/moved we will not notice and not rerun gazelle.
+  _exec(repo_ctx, ["touch", autogo])
+  # Run the autogo binary (see shadow.go), which clones the dir tree and symlinks in relevant files
+  _exec(repo_ctx, [autogo, root_workspace, repo_ctx.path(".")])
+  # Run gazelle to auto-create rules
+  _exec(repo_ctx,
+    [gazelle, "--repo_root", repo_ctx.path(""), "--go_prefix", repo_ctx.attr.prefix])
+
+_autogo_generate = repository_rule(
+    implementation=_impl,
+    local=True,
+    attrs={
+        "prefix": attr.string(mandatory=True),
+        "_workspace": attr.label(
+            allow_single_file=True,
+            default="@//:WORKSPACE"),
+        "_autogo_linux": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/linux_amd64/autogo"),
+        "_autogo_darwin": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/darwin_amd64/autogo"),
+        "_gazelle_linux": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/linux_amd64/gazelle"),
+        "_gazelle_darwin": attr.label(
+            allow_single_file=True,
+            default="//autogo:bin/darwin_amd64/gazelle"),
+        },
+)
+
+def autogo_generate(name="autogo", **kw):
+  """Generates automatic rules for go packages at @name (usually @autogo).
+
+  Usage:
+    autogo_generate(name="autogo", prefix="github.com/my/go/get/path")
+  """
+  go_rules_dependencies()
```

Comment:
we probably want some way to specify go version and/or custom go toolchains.

Reply:
that seems slightly important?

Reply:
Not really. I'm not interested in designing "I want to express my complicated go requirements as arguments to `autogo_generate()`" right now.

People who have complicated requirements can call `go_rules_dependencies()` and `go_register_toolchains()` in whatever manner they want before calling `autogo_generate()`, and bazel will setup whatever they specify.

The code here does nothing in this scenario as the rules are already defined: 
* https://github.com/bazelbuild/rules_go/blob/master/go/private/repositories.bzl#L147
* https://github.com/bazelbuild/rules_go/blob/master/go/toolchain/toolchains.bzl#L267

So I really don't see the need to design that now, when we can just do it later.

What are you concerned about?

Reply:
er if we don't plumb through the args here then won't we have different toolchains registered than elsewhere in the workspace potentially? or are you envisioning that this will be the only caller? either way we depend pretty heavily on registering a specific pinned go version

Reply:
No, we will wind up with the same values.

The first call to `go_rules_dependencies()` will load all the `_maybe()` git repos (unless something even earlier already registered them). Subsequent calls will skip this because the repo already exists.

The first call to `go_register_toolchains()` will define the `go_sdk` repo -- subsequent calls will skip this because the repo already exists.

Would you be happier if I put this `if name not in native.existing_rules()` block?

Reply:
I think maybe we just should note that you can select the toolchains by making sure you load them before this? No reason to duplicate the work rules_go already does then.